File: extension/src/shared/normalize.ts
/**
 * Text normalization. Two functions, two different jobs.
 *
 * Port of training/src/slopmarker/data/normalize.py. Pinned by fixtures/normalize.json;
 * both test suites read that file. If you disagree with a choice here, change the Python
 * side and regenerate the fixture -- do not reinterpret it.
 *
 * `collapseWhitespace` is what the *model* sees: it only regularizes whitespace, so curly
 * quotes and em dashes survive. Those are real signal for AI text.
 *
 * `normalizeForHash` is what the *score cache key* is built from (scope.md 7.4). It folds
 * quotes, dashes and invisibles too, so syndicated text differing only in typography is
 * scored once.
 *
 * Every character class is an explicit codepoint set and no regex `\s` appears anywhere:
 * Python's `\s` matches U+001C-U+001F, JavaScript's matches U+FEFF and not those. Relying
 * on either would guarantee a drift bug that only fires on scraped web text.
 */

/** Iterate codepoints, not UTF-16 code units. Python's `for ch in text` does the same. */
export function toCodePoints(text: string): string[] {
  return Array.from(text);
}

/** Every character we treat as whitespace. Nothing else counts. */
export const WHITESPACE: ReadonlySet<string> = new Set(
  [
    0x09, // tab
    0x0a, // line feed
    0x0b, // vertical tab
    0x0c, // form feed
    0x0d, // carriage return
    0x20, // space
    0x00a0, // no-break space
    0x1680, // ogham space mark
    // en quad .. hair space (0x2000..0x200A); note 0x200B is INVISIBLE, not whitespace
    0x2000, 0x2001, 0x2002, 0x2003, 0x2004, 0x2005, 0x2006, 0x2007, 0x2008, 0x2009, 0x200a,
    0x2028, // line separator
    0x2029, // paragraph separator
    0x202f, // narrow no-break space
    0x205f, // medium mathematical space
    0x3000, // ideographic space
  ].map((cp) => String.fromCodePoint(cp)),
);

/**
 * Zero-width and format characters deleted outright. U+200C ZWNJ and U+200D ZWJ are
 * deliberately NOT here: they are load-bearing in emoji sequences and in Indic and Arabic
 * scripts, and deleting them would mangle real text.
 */
export const INVISIBLE: ReadonlySet<string> = new Set(
  [
    0x00ad, // soft hyphen
    0x200b, // zero-width space
    0x2060, // word joiner
    0xfeff, // zero-width no-break space / BOM
  ].map((cp) => String.fromCodePoint(cp)),
);

const SINGLE_QUOTES = [0x2018, 0x2019, 0x201a, 0x201b, 0x2032]; // curly quotes, prime
const DOUBLE_QUOTES = [0x201c, 0x201d, 0x201e, 0x201f, 0x2033]; // curly quotes, double prime
const DASHES = [0x2010, 0x2011, 0x2012, 0x2013, 0x2014, 0x2015, 0x2212]; // hyphen .. minus

export const ELLIPSIS = String.fromCodePoint(0x2026);

const FOLD: ReadonlyMap<string, string> = new Map([
  ...SINGLE_QUOTES.map((cp) => [String.fromCodePoint(cp), "'"] as const),
  ...DOUBLE_QUOTES.map((cp) => [String.fromCodePoint(cp), '"'] as const),
  ...DASHES.map((cp) => [String.fromCodePoint(cp), "-"] as const),
]);

/** Map every whitespace character to a space, collapse runs, strip the ends. */
export function collapseWhitespace(text: string): string {
  const out: string[] = [];
  let pendingSpace = false;
  for (const ch of text) {
    if (WHITESPACE.has(ch)) {
      pendingSpace = out.length > 0; // leading whitespace never sets it
    } else {
      if (pendingSpace) {
        out.push(" ");
        pendingSpace = false;
      }
      out.push(ch);
    }
  }
  return out.join("");
}

/** Aggressive normalization used only to build the score-cache key. */
export function normalizeForHash(text: string): string {
  const out: string[] = [];
  for (const ch of text.normalize("NFC")) {
    if (INVISIBLE.has(ch)) continue;
    if (ch === ELLIPSIS) {
      out.push("...");
    } else {
      out.push(FOLD.get(ch) ?? ch);
    }
  }
  return collapseWhitespace(out.join(""));
}

/**
 * sha256 of the normalized form, hex. The score-cache key of scope.md 7.4.
 *
 * Async because it uses Web Crypto, which is what both the content script and the host
 * have. Node 18+ exposes the same `crypto.subtle`, so the fixture test runs unchanged.
 */
export async function contentHash(text: string): Promise<string> {
  const bytes = new TextEncoder().encode(normalizeForHash(text));
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/**
 * Word count, defined on whitespace-collapsed text so it needs no regex at all.
 * This is the one definition used by training, by windowing and by the extension.
 */
export function countWords(text: string): number {
  const collapsed = collapseWhitespace(text);
  return collapsed ? collapsed.split(" ").length : 0;
}

File: training/src/slopmarker/data/normalize.py
r"""Text normalization. Two functions, two different jobs.

`collapse_whitespace` is what the *model* sees: it only regularizes whitespace, so
curly quotes and em dashes survive. Those are real signal for AI text and stripping
them would throw the signal away.

`normalize_for_hash` is what the *score cache key* is built from (scope.md 7.4). It
folds quotes, dashes and invisibles too, so that syndicated text differing only in
typography is scored once.

Both are ported to TypeScript and pinned by fixtures/normalize.json. Every character
class is written out as an explicit codepoint and no regex `\s` is used anywhere:
Python's `\s` matches U+001C-U+001F, JavaScript's matches U+FEFF and not those.
Relying on it would guarantee a drift bug that only fires on scraped web text.
"""

from __future__ import annotations

import hashlib
import re
import unicodedata

# Every character we treat as whitespace. Nothing else counts.
WHITESPACE = frozenset(
    map(
        chr,
        (
            0x09,  # tab
            0x0A,  # line feed
            0x0B,  # vertical tab
            0x0C,  # form feed
            0x0D,  # carriage return
            0x20,  # space
            0x00A0,  # no-break space
            0x1680,  # ogham space mark
            *range(0x2000, 0x200B),  # en quad .. hair space
            0x2028,  # line separator
            0x2029,  # paragraph separator
            0x202F,  # narrow no-break space
            0x205F,  # medium mathematical space
            0x3000,  # ideographic space
        ),
    )
)

# Zero-width and format characters deleted outright. U+200C ZWNJ and U+200D ZWJ are
# deliberately NOT here: they are load-bearing in emoji sequences and in Indic and
# Arabic scripts, and deleting them would mangle real text.
INVISIBLE = frozenset(
    map(
        chr,
        (
            0x00AD,  # soft hyphen
            0x200B,  # zero-width space
            0x2060,  # word joiner
            0xFEFF,  # zero-width no-break space / BOM
        ),
    )
)

_SINGLE_QUOTES = (0x2018, 0x2019, 0x201A, 0x201B, 0x2032)  # curly quotes, prime
_DOUBLE_QUOTES = (0x201C, 0x201D, 0x201E, 0x201F, 0x2033)  # curly quotes, double prime
_DASHES = (0x2010, 0x2011, 0x2012, 0x2013, 0x2014, 0x2015, 0x2212)  # hyphen .. minus
ELLIPSIS = chr(0x2026)

_FOLD: dict[str, str] = {
    **{chr(cp): "'" for cp in _SINGLE_QUOTES},
    **{chr(cp): '"' for cp in _DOUBLE_QUOTES},
    **{chr(cp): "-" for cp in _DASHES},
}


def collapse_whitespace(text: str) -> str:
    """Map every whitespace character to a space, collapse runs, strip the ends."""
    out: list[str] = []
    pending_space = False
    for ch in text:
        if ch in WHITESPACE:
            pending_space = bool(out)
        else:
            if pending_space:
                out.append(" ")
                pending_space = False
            out.append(ch)
    return "".join(out)


def normalize_for_hash(text: str) -> str:
    """Aggressive normalization used only to build the score-cache key."""
    text = unicodedata.normalize("NFC", text)
    out: list[str] = []
    for ch in text:
        if ch in INVISIBLE:
            continue
        if ch == ELLIPSIS:
            out.append("...")
        else:
            out.append(_FOLD.get(ch, ch))
    return collapse_whitespace("".join(out))


def content_hash(text: str) -> str:
    return hashlib.sha256(normalize_for_hash(text).encode("utf-8")).hexdigest()


def count_words(text: str) -> int:
    """Word count, defined on whitespace-collapsed text so it needs no regex at all.

    This is the one definition used by training, by windowing and by the extension.
    """
    collapsed = collapse_whitespace(text)
    return len(collapsed.split(" ")) if collapsed else 0


# Markdown syntax, removed from both classes before windowing.
#
# This is not "removing model style". At inference the extension reads rendered DOM
# text: a heading arrives as the words inside an <h2>, and bold arrives as the words
# inside a <strong>. The literal `#` and `**` characters cannot reach the model in
# production. Human web text is extracted the same way and so contains none of them,
# which makes a literal asterisk an almost perfect classifier -- measured at 8x more
# frequent in AI text than human before this ran.
#
# The *choice* to emphasise or to use headings survives, as word choice and as
# structure. Only the syntax that production never sees is removed.
_MD_HEADING = re.compile(r"^[ \t]{0,3}#{1,6}[ \t]+", re.M)
_MD_EMPHASIS = re.compile(r"(\*{1,3}|_{1,3})(?=\S)(.+?)(?<=\S)\1", re.S)
_MD_BULLET = re.compile(r"^[ \t]{0,3}([-*+])[ \t]+", re.M)
_MD_RULE = re.compile(r"^[ \t]{0,3}([-*_])(?:[ \t]*\1){2,}[ \t]*$", re.M)
_MD_CODE_FENCE = re.compile(r"^[ \t]{0,3}```.*$", re.M)
_MD_INLINE_CODE = re.compile(r"`([^`\n]+)`")
_MD_LINK = re.compile(r"\[([^\]\n]+)\]\([^)\n]*\)")
_MD_BLOCKQUOTE = re.compile(r"^[ \t]{0,3}>[ \t]?", re.M)


def strip_markdown(text: str) -> str:
    """Remove markdown syntax, keeping the words it wrapped."""
    text = _MD_CODE_FENCE.sub("", text)
    text = _MD_RULE.sub("", text)
    text = _MD_LINK.sub(r"\1", text)
    text = _MD_INLINE_CODE.sub(r"\1", text)
    text = _MD_EMPHASIS.sub(r"\2", text)
    text = _MD_HEADING.sub("", text)
    text = _MD_BULLET.sub("", text)
    return _MD_BLOCKQUOTE.sub("", text)
